### ISSUE TITLE
Dynamic styles break getSnapshotBeforeUpdate

### ISSUE THREAD

**Issue by mattgperry** [vercel/styled-jsx#671]
#### Do you want to request a *feature* or report a *bug*?

Bug

#### What is the current behavior?

If we have a component that sets styles dynamically:

```jsx
<style jsx>{`
  div {
    width: ${isBig ? "200px" : "100px"}
  }
`}</style>
```

And try to measure that using `getSnapshotBeforeUpdate`, the measurements of the element are invariably incorrect. `height` is usually `0` and `width` the size of some ancestor.

This breaks Framer Motion's layout animations.

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

https://codesandbox.io/s/flamboyant-sunset-l9p15?file=/pages/index.js

#### What is the expected behavior?

The returned measurements should match the observed width/height. The fact it doesn't makes me think that the application of dynamic styles is also triggering two layouts rather than one, as there must be some "in between" time after the old styles are removed and the new ones are added.

#### Environment (include versions)

* OS: MacOS 10.15.3
* Browser: Chrome 85
* styled-jsx (version): Whatever's shipped with `next` 9.5.5

#### Did this work in previous versions?

Don't know

**Comment by giuseppeg** [comment]
> getSnapshotBeforeUpdate() is invoked right before the most recently rendered output is committed to e.g. the DOM

shouldn't you measure in `componentDidUpdate`?

**Comment by giuseppeg** [comment]
anyway I assume that the issue is with styled-jsx injecting and removing styles during rendering (see why - 1 - https://github.com/vercel/styled-jsx/pull/484#issuecomment-584474687) https://github.com/vercel/styled-jsx/blob/d07f76ed4952a7f0f2c3d6401d2b6667661b0a9a/src/style.js#L36-L43

I wonder if there is a solution to - 1 -

Did you try to track the rect in componentDidMount/Update?

**Comment by mattgperry** [comment]
The purpose of `getSnapshotBeforeUpdate` to record `DOMRect` **before** a React update. This can then be compared against another measurement in `componentDidUpdate`.

**Comment by giuseppeg** [comment]
yeah I figured you wanted to measure before the update:

> But I guess you are talking about the previous rect right?

I am afraid that the issue is due to the fact that we unmount the previous styles in the render phase (L42) - here's a reproduction https://codesandbox.io/s/sleepy-lumiere-cykk3?file=/pages/index.js

1. we remove the old styles
2. the `div` in the dom has still the old scoped class so basically it is unstyled
3. `getSnapshotBeforeUpdate` runs and measures an unstyled element
4. React commits the changes to the DOM
5. the `div` has now the new scoped classname

Ideally we should remove and inject styles in `componentDidUpdate` but when the styles contain animations/transitions and a parent component triggers layout (quoting @sebmarkbage):

> This causes Chrome to think this is an update to an already visible component and triggers a transition

This is the reason why styled-jsx component is the way it is now:

https://github.com/vercel/styled-jsx/blob/d07f76ed4952a7f0f2c3d6401d2b6667661b0a9a/src/style.js#L36-L46

with this ugly hack the styles are in the document before React commits and this won't trigger transitions but it might cause issues like the one you reported.

**Comment by huozhi** [comment]
This is resolved by styled-jsx v5 beta and latest next canary
